core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -41,306 +40,341 @@ object ConsumerGroupCommand {
+    val configsToBeAdded = opts.options.valuesOf(opts.configOpt).asScala.map(_.split("""\s*=\s*"""))
+    require(configsToBeAdded.forall(config => config.length == 2),
+      "Invalid config: all configs to be added must be in the format \"key=val\".")
+    val props = new Properties
+    configsToBeAdded.foreach(pair => props.setProperty(pair(0).trim, pair(1).trim))
+    props
   }
 
-  def createAndGetAdminClient(opts: ConsumerGroupCommandOptions): AdminClient = {
-    AdminClient.createSimplePlaintext(opts.options.valueOf(opts.bootstrapServerOpt))
-  }
+  sealed trait ConsumerGroupService {
 
-  def describe(zkUtils: ZkUtils, opts: ConsumerGroupCommandOptions) {
-    val useNewConsumer = opts.options.has(opts.newConsumerOpt)
-    val group = opts.options.valueOf(opts.groupOpt)
-    val configs = parseConfigs(opts)
-    val channelSocketTimeoutMs = configs.getProperty("channelSocketTimeoutMs", "600").toInt
-    val channelRetryBackoffMs = configs.getProperty("channelRetryBackoffMsOpt", "300").toInt
+    def list(): Unit
 
-    println("%s, %s, %s, %s, %s, %s, %s"
-      .format("GROUP", "TOPIC", "PARTITION", "CURRENT OFFSET", "LOG END OFFSET", "LAG", "OWNER"))
+    def describe() {
+      val group = opts.options.valueOf(opts.groupOpt)
+      println("GROUP, TOPIC, PARTITION, CURRENT OFFSET, LOG END OFFSET, LAG, OWNER")
+      describeGroup(group)
+    }
 
-    if (!useNewConsumer) {
-      val topics = zkUtils.getTopicsByConsumerGroup(group)
-      if (topics.isEmpty) {
-        println("No topic available for consumer group provided")
-      } else {
-        topics.foreach(topic => describeTopic(zkUtils, group, topic, channelSocketTimeoutMs, channelRetryBackoffMs, opts))
-      }
-    } else {
-      val consumers = createAndGetAdminClient(opts).describeConsumerGroup(group)
+    def close(): Unit
 
-      if (consumers.isEmpty) {
-        println(s"Consumer group, ${group}, does not exist or is rebalancing.")
-      } else {
-        consumers.foreach(x => describeTopicPartition(zkUtils, group, channelSocketTimeoutMs, channelRetryBackoffMs, opts, x.assignment.map(tp => new TopicAndPartition(tp.topic(), tp.partition())), Option("%s_%s".format(x.clientId, x.clientHost))))
+    protected def opts: ConsumerGroupCommandOptions
+
+    protected def getLogEndOffset(topic: String, partition: Int): LogEndOffsetResult
+
+    protected def describeGroup(group: String): Unit
+
+    protected def describeTopicPartition(group: String,
+                                         topicPartitions: Seq[TopicAndPartition],
+                                         getPartitionOffset: TopicAndPartition => Option[Long],
+                                         getOwner: TopicAndPartition => Option[String]): Unit = {
+      topicPartitions
+        .sortBy { case topicPartition => topicPartition.partition }
+        .foreach { topicPartition =>
+          describePartition(group, topicPartition.topic, topicPartition.partition, getPartitionOffset(topicPartition),
```

Comment:
A little annoying that we end up constructing a new KafkaConsumer for every partition. Every time we create it, we have to incur the overhead of looking up the coordinator.